core/iwasm/aot/aot_loader.c

```diff
@@ -1119,11 +1131,142 @@ destroy_table_init_data_list(AOTTableInitData **data_list, uint32 count)
 {
     uint32 i;
     for (i = 0; i < count; i++)
-        if (data_list[i])
+        if (data_list[i]) {
+#if WASM_ENABLE_GC != 0
+            uint32 j;
+            for (j = 0; j < data_list[i]->value_count; j++) {
+                destroy_init_expr(&data_list[i]->init_values[j]);
+            }
+#endif
             wasm_runtime_free(data_list[i]);
+        }
     wasm_runtime_free(data_list);
 }
 
+#if WASM_ENABLE_GC != 0
+static bool
+load_init_expr(const uint8 **p_buf, const uint8 *buf_end, AOTModule *module,
+               InitializerExpression *expr, char *error_buf,
+               uint32 error_buf_size)
+{
+    const uint8 *buf = *p_buf;
+    uint32 init_expr_type = 0;
+    bool free_if_fail = false;
+
+    buf = (uint8 *)align_ptr(buf, 4);
+
+    read_uint32(buf, buf_end, init_expr_type);
+
+    switch (init_expr_type) {
+        case INIT_EXPR_NONE:
+            break;
+        case INIT_EXPR_TYPE_I32_CONST:
+        case INIT_EXPR_TYPE_F32_CONST:
+            read_uint32(buf, buf_end, expr->u.i32);
+            break;
+        case INIT_EXPR_TYPE_I64_CONST:
+        case INIT_EXPR_TYPE_F64_CONST:
+            read_uint64(buf, buf_end, expr->u.i64);
+            break;
+        case INIT_EXPR_TYPE_V128_CONST:
+            read_byte_array(buf, buf_end, &expr->u.v128, sizeof(expr->u.v128));
+            break;
+        case INIT_EXPR_TYPE_GET_GLOBAL:
+            read_uint32(buf, buf_end, expr->u.global_index);
+            break;
+        case INIT_EXPR_TYPE_REFNULL_CONST:
+            read_uint32(buf, buf_end, expr->u.type_index);
+            break;
+        case INIT_EXPR_TYPE_FUNCREF_CONST:
+            read_uint32(buf, buf_end, expr->u.ref_index);
+            break;
+        case INIT_EXPR_TYPE_I31_NEW:
+            read_uint32(buf, buf_end, expr->u.i32);
+            break;
+        case INIT_EXPR_TYPE_STRUCT_NEW_CANON:
+        {
+            uint64 size;
+            uint32 type_idx, field_count;
+            WASMStructNewInitValues *init_values = NULL;
+
+            read_uint32(buf, buf_end, type_idx);
+            read_uint32(buf, buf_end, field_count);
+
+            size = offsetof(WASMStructNewInitValues, fields)
+                   + sizeof(WASMValue) * (uint64)field_count;
+            if (!(init_values =
+                      loader_malloc(size, error_buf, error_buf_size))) {
+                return false;
+            }
+            free_if_fail = true;
+
+            if (field_count > 0) {
+                init_values->type_idx = type_idx;
+                init_values->count = field_count;
+
+                size = sizeof(WASMValue) * (uint64)field_count;
+                read_byte_array(buf, buf_end, init_values->fields, size);
+            }
+
+            expr->u.data = init_values;
```

Comment:
Should move this line to be after L1201? or in L1208, if read_byte_array fails, `wasm_runtime_free(expr->u.data)` may free an invalid memory.